### ISSUE TITLE
Lambda capture types are unnecessarily conservative

### ISSUE THREAD

**Issue by Timmmm** [microsoft/pyright#2335]
Consider the following code:

```python
def foo(x: Optional[int]) -> int:
    if x is None:
        x = 5
    c = lambda: x
    return c()
```

Pyright gives an error on the last line, because `x` in the lambda is inferred to be `int | None`. But at the point where `c` is defined, and `x` is captured, Pyright has already narrowed `x` to just `int`.

Shouldn't the lambda captures take the types of the variables they capture from the point where they are defined, including narrowing? Or am I missing some devious Python behaviour that makes this not safe?

**Comment by erictraut** [comment]
Narrowed types are not safe to use for variables that are captured and used in a different execution context (a module, function or lambda). That's because a type checker validates each execution context independently, and it does not (and generally cannot) know the order of invocation.

Consider the following:
```python
def foo(x: Optional[int]) -> int:
    if x is None:
        x = 5
    c = lambda: x
    x = None
    return c()
```

The recommended approach is to use a new local variable that is captured by the inner execution scope. Its type can be determined without type narrowing or assumptions about execution order.

```python
def foo(x: Optional[int]) -> int:
    if x is None:
        x = 5
    y = x
    c = lambda: y
    return c()
```

**Comment by Timmmm** [comment]
Ah of course! Thanks!

**Comment by erictraut** [comment]
After thinking about this more, I found a way to safely handle this case in a safe manner. We can apply type narrowing to a local variable or parameter that is captured by an inner-scoped lambda or function in cases where the type analyzer can prove that there is no code path that assigns a new value to the symbol after the lambda or def statement.

This will be included in the next release of pyright and pylance.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.172, which I just published. It will also be in the next release of pylance.

**Comment by Timmmm** [comment]
Amazing! Gotta say you're killing it with Pyright. Ask for a raise. :-)
